ambry-commons/src/main/java/com/github/ambry/compression/CompressionFactory.java

```diff
@@ -0,0 +1,124 @@
+
+    synchronized(CompressionFactory.class) {
+      if (defaultFactory == null) {
+        defaultFactory = new CompressionFactory();
+
+        // ****** WARNING ********
+        // You may register/add more compression services, but DO NOT remove any existing compression service.
+        // Decompression relies on this map to locate the decompression algorithm using the algorithm name.
+        defaultFactory.register(new LZ4Compression());
+        defaultFactory.register(new ZstdCompression());
+      }
+    }
+
+    return defaultFactory;
+  }
+
+  // Map from algorithm name to instance of compression service.
+  private final Map<String, Compression> registeredCompressionMap = new HashMap<>();
+
+  // The default compressor to use to compress new data.
+  private Compression defaultCompressor;
+
+  /**
+   * Register a compression service.  Every compression service has a unique name.
+   * Decompression will use the registered compression services to find the decompression algorithm.
+   *
+   * @param newCompressionService The compression service to register.
+   */
+  public void register(Compression newCompressionService) {
+    if (newCompressionService == null) throw new IllegalArgumentException("newCompressionService");
+    registeredCompressionMap.put(newCompressionService.getAlgorithmName(), newCompressionService);
+  }
+
+  /**
+   * To get a compressor by algorithm name.
+   * The application config could call this method to set the default compressor algorithm.
+   *      CompressionFactory factory = CompressionFactory.getDefault();
+   *      Compression compressor = factory.getCompressorByName("LZ4");
+   *      factory.setDefaultCompressor(compressor);
+   *
+   * @param algorithmName Name of the algorithm to get.
+   * @return Instance of the compression algorithm; null if not found.
+   */
+  public Compression getCompressorByName(String algorithmName) {
+    return registeredCompressionMap.get(algorithmName);
+  }
+
+  /**
+   * Set the default compressor to use when compressing data.
+   * @param newCompressor The new compressor to use.
+   */
+  public void setDefaultCompressor(Compression newCompressor) {
+    defaultCompressor = newCompressor;
+  }
+
+  /**
+   * Get the default compressor to use when compressing data.
+   * It returns the compressor set by setDefaultCompressor().
+   * @return The default compressor.  Return null if not set.
+   */
+  public Compression getDefaultCompressor() {
+    return defaultCompressor;
+  }
+
+  /**
+   * Given the compressed buffer generated by compressor's compress() method,
+   * find the decompressor based on the algorithm name stored in compressedBuffer.
+   *
+   * @param compressedBuffer The compressed buffer generated by the compress() method.
+   * @return The decompressor that can decompress this compressed buffer.  Null if no decompressor found.
+   */
+  public Compression getDecompressor(byte[] compressedBuffer) {
```

Comment:
I don't think we differentiate Compressor from Decompressor. So it's probably a bit confusing as an interface. Maybe change the name to `getCompressionFromCompressedData`, or something short.